src/transport/manager/mod.rs

```diff
@@ -724,6 +724,49 @@ impl TransportManager {
         Ok(())
     }
 
+    // Update the address on a dial failure.
+    fn update_address_on_dial_failure(&mut self, mut address: Multiaddr, error: &DialError) {
+        let mut peers = self.peers.write();
+
+        let score = AddressStore::error_score(error);
+
+        // Check if the address corresponds to a different peer ID than the one we're
+        // dialing. This can happen if the node operation restarts the node.
+        //
+        // In this case the address is reachable, however the peer ID is different.
+        // Keep track of this address for future dials.
+        //
+        // Note: this is happening quite often in practice and is the primary reason
+        if let DialError::NegotiationError(NegotiationError::PeerIdMismatch(_, provided)) = error {
+            let context = peers.entry(*provided).or_insert_with(|| PeerContext::default());
+
+            if !std::matches!(address.iter().last(), Some(Protocol::P2p(_))) {
+                address.pop();
+            }
+            context
+                .addresses
+                .insert(AddressRecord::new(&provided, address.clone(), score, None));
```

Comment:
May be I don't understand something, but not sure if this is a good idea:
1. If we don't discover the "new" `PeerId` through Kademlia, we won't dial this address anyway.
2. Is it possible to generate `PeerIdMismatch` with arbitrary `PeerId` without cryptographically proving you own the private key? Then malicious actor can poison the address store of arbitrary peer by joining the DHT and advertising random `PeerId`, but on connection from the poisoned node provide the victim's `PeerId` via `PeerIdMismatch`.